Add release notes for meshlet BVH culling (#20526)
bevyengine/bevy: release-content/release-notes/meshlet-bvh-culling.md

---

> title: Meshlet BVH culling

Reviewer: ```suggestion
title: Virtual Geometry BVH culling
```

I'm trying to avoid calling it meshlets. We should really do the rename soon...

Author: Hmm, yeah, probably will have to wait for 0.18 though.

authors: ["@SparkyPotato", "@atlv24"]
pull_requests: [19318]
---

(TODO: Embed example screenshot here)

Bevy's virtual geometry has been greatly optimized with BVH-based culling, leading to almost true scene-complexity invariance on the GPU.

> This gets rid of the previous cluster limit that limited the world to 2^24 clusters (about 4 billion triangles).

Reviewer: ```suggestion
This also gets rid of the previous cluster limit that limited the world to 2^24 clusters (about 4 billion triangles).
```

> There are now *no* hardcoded limits to scene size, only unique instance limits due to VRAM usage (since streaming is not yet implemented),
> and total instance limits due the current architecture requiring all instances to be uploaded to the GPU every frame.

Reviewer: ```suggestion
There are now *no* hardcoded limits to scene size. In practice you will only be limited by asset VRAM usage (since streaming is not yet implemented),
and total instance count due the current code requiring all instances to be re-uploaded to the GPU every frame.
```

The screenshot above has 130,000 dragons in the scene, each with about 870,000 triangles, leading to over *115 billion* total triangles in the scene.

> However, this still runs at 60 fps on an RTX 4070 at 1440p, with most of the time being due to the instance upload CPU bottleneck mentioned above (taking 14 ms of CPU time).

Reviewer: I don't think this sentence would mean much to people not very familiar with the rendering code.

> Speaking of GPU cost, the scene above renders in about 3.5 ms on the 4070, with ~3.1 ms being spent on the geometry render and ~0.4 ms on the material evaluation.

Author: I was referring to the 4070 + 1440p combo mentioned above

> Increasing the instance count to over 1 million (almost *900 billion triangles*!), the total increase to about 4.5 ms, with ~4.1 ms on geometry render and material evaluation remaining constant at ~0.4 ms.

Reviewer 2: ```suggestion
After increasing the instance count to over 1 million (almost *900 billion triangles*!), the GPU time increases to about 4.5 ms, with ~4.1 ms on geometry render and material evaluation remaining constant at ~0.4 ms.
```

> After increasing the instance count to over 1 million (almost *900 billion triangles*!), the GPU time increases to about 4.5 ms, with ~4.1 ms on geometry render and material evaluation remaining constant at ~0.4 ms.

Reviewer: Add something like "this is X times the triangle count, but only Y more GPU time"

---
title: Virtual Geometry BVH culling
authors: ["@SparkyPotato", "@atlv24"]
pull_requests: [19318]
---

(TODO: Embed example screenshot here)

Bevy's virtual geometry has been greatly optimized with BVH-based culling, making the cost of rendering nearly independent of scene geometry.

> Comparing the sample scene show above with 130k dragon instances to one with over 1 million instances, total GPU rendering time only increases by 30%.

Reviewer 2: ```suggestion
Comparing the sample scene shown above with 130k dragon instances to one with over 1 million instances, total GPU rendering time only increases by 30%.
```

This also gets rid of the previous cluster limit that limited the world to 2^24 clusters (about 4 billion triangles).
There are now *no* hardcoded limits to scene size. In practice you will only be limited by asset VRAM usage (since streaming is not yet implemented),
and total instance count due the current code requiring all instances to be re-uploaded to the GPU every frame.

The screenshot above has 130,000 dragons in the scene, each with about 870,000 triangles, leading to over *115 billion* total triangles in the scene.

Speaking of concrete GPU cost, the scene above renders in about 3.5 ms on the 4070, with \~3.1 ms being spent on the geometry render and \~0.4 ms on the material evaluation.

> After increasing the instance count to over 1 million (almost *900 billion triangles*!), the total increase to about 4.5 ms, with \~4.1 ms on geometry render and material evaluation remaining constant at ~0.4 ms.

Reviewer 2: ```suggestion
After increasing the instance count to over 1 million (almost *900 billion triangles*!), the total increases to about 4.5 ms, with \~4.1 ms on geometry render and material evaluation remaining constant at ~0.4 ms.
```

---
title: Virtual Geometry BVH culling
authors: ["@SparkyPotato", "@atlv24"]
pull_requests: [19318]
---

(TODO: Embed example screenshot here)

Bevy's virtual geometry has been greatly optimized with BVH-based culling, making the cost of rendering nearly independent of scene geometry.
Comparing the sample scene show above with 130k dragon instances to one with over 1 million instances, total GPU rendering time only increases by 30%.

This also gets rid of the previous cluster limit that limited the world to 2^24 clusters (about 4 billion triangles).
There are now *no* hardcoded limits to scene size. In practice you will only be limited by asset VRAM usage (since streaming is not yet implemented),
and total instance count due the current code requiring all instances to be re-uploaded to the GPU every frame.

The screenshot above has 130,000 dragons in the scene, each with about 870,000 triangles, leading to over *115 billion* total triangles in the scene.

> However, this still runs at 60 fps on an RTX 4070 at 1440p, but only due to the instance re-upload mentioned above.

Reviewer: I would probably remove this section, it doesn't add anything
